### DISCUSSION TITLE
Potential Improvements to PDF Export

### DISCUSSION THREAD

**Discussion by lloy0076** [code4recovery/12-step-meeting-list#430] [Ideas]
The current behaviour of the export to PDF is a little odd in several ways:

**Meeting types run into the address**

[meeting-schedule.pdf](https://github.com/code4recovery/12-step-meeting-list/files/6794168/meeting-schedule.pdf)

Notice how in the above the meeting types run into the address.

**Online Meetings**

You'd not know they're online meetings! So Nooners during the week are all online but it's listed as meeting at Maple St!

**Possible Changes**

[meeting-schedule(14).pdf](https://github.com/code4recovery/12-step-meeting-list/files/6794171/meeting-schedule.14.pdf)

So I played about a bit with the code and came up with the above; notice that:

- I've put the meeting types onto their own line
- If it's an online meeting I've made it just print out the URL

**Further Changes/Enhancements**

1. Make it more configurable
2. Make it template-able
3. Add internationalisation (i18n) --> it's very English specific ATM
4. Print out the name of the start day

(If you change L154 https://github.com/code4recovery/12-step-meeting-list/blob/d06856d27e52fddb93cd7dfa69bacc805886f42e/includes/pdf.php#L150 to something like):

```
if (isset($meeting['attendance_option']) && $meeting['attendance_option'] === 'in_person') {

		$this->Cell($day_width, .1, '', 0, 0);
	        $this->Cell($time_width, .1, '', 0, 0);
                $this->MultiCell($right_width, .1, $location . ', ' . @$meeting['formatted_address'], 0, 'L');
                $this->Ln(.1);
} else {
		$this->Cell($day_width, .1, '', 0, 0);
	        $this->Cell($time_width, .1, '', 0, 0);
                $this->MultiCell($right_width, .1, 'Check this URL: ' . $meeting['url'], 0, 'L');
		$this->Ln(.1);
}
```

Now, I might be motivated to put some effort into this but am wondering if there is a ROADMAP to improve this functionality already, who else might be working on this, or what any work in this area might affect.

**Comment by tech2serve** [comment]
Moving this to our new Discussions page.

**Comment by lloy0076** [comment]
It appears that [Chris T's plugin](https://github.com/cdtoews/12-step-meeting-pdf) seems to work (give or take PHP warnings).

I wonder if the group here would prefer such functionality to be provided by an 'outside' plugin, if there's a possibility of collaboration and so froth?

**Comment by tech2serve** [comment]
HI @lloy0076 --

Our dev team needs to discuss the future of PDF support. Right now there are handful of options:

1. Revamp the code in place. 
2. Use a separate plugin (Could be Chris T's plugin or something created anew) or service.
3. Drop support.

The latter option may seem like overkill, but some of us just aren't convinced AA service entities should continue to print off meeting lists that are not accurate, often not used, and don't reflect the changing times. 

I know right off that a number of people will tar and feather me for the suggestion :-) But even the manager of a CSO I support, a man who is a self-proclaimed technology neophyte, indicated he is considering discontinuing the printing and distribution of paper meeting lists. 

All that said, I suspect option 3 isn't realistic yet, so I believe option 2 becomes a better choice. Chris has not updated his plugin for 10 months, but I do think it still provides more capability than the PDF printing feature in TSML. 

If you're looking for coding opportunities, perhaps you can serve as a bridge between the two projects?

**Reply by lloy0076** [reply]
Interestingly, our district registrar mentioned that a meeting attendee just last night (9 July) had a *paper* meeting list in theirhands (old but still somewhat up to date) which was how the attendee found the meeting.

I can see a collaboration between this and a "sibling project" could work.

**Reply by joshreisner** [reply]
I think a microservice architecture could be an option here. How that would work:

* set up a new hosted service at `https://pdf.code4recovery.org` (or another domain)
* this can be running PHP / Laravel / Python / Node / maintainer's choice
* to generate a PDF on the fly, go to a route on this domain that sends the JSON feed as a parameter, eg `/generate?json=https://path/to/my/meeting.json`
* the microservice grabs this JSON, and returns a PDF
* we add links inside TSML to this service

Benefits of this architecture:
* ability to install whatever platform / libraries you want
* updates and support aren't tied to TSML
* update once / update everywhere
* no need to support various versions of PHP or other hosting oddities

**Reply by lloy0076** [reply]
@joshreisner Another benefit of such an architecture is that one could move away from WordPress (or even PHP!!!) if one decided that WordPress plugins are just a little too annoying to work with -- at least on server side.

On the other hand, someone / somewhere would need to host such a service; I don't know how well used it would be but it would need to be somewhat secure (if they're going to Zoom Bomb meetings, why not hijack the meeting PDF generator to print off topic messages), and somewhat resilient to high use [the good case] or concerted DDOS [the bad case].

The other thought I have -- of course -- is that once we're in printed meeting pamphlets, we get into the whole "what should such a thing look like" and discover various things such as:

* Why typesetters have jobs;
* Why graphic designers (who work with typesetters) have jobs;
* How to please many people.

So, District 42 has the classic three fold. Others might have a single page. Some might prefer the logo to be 25px, the font to be Comic Sans [yuck!], and a legend printed...then an AA group in a LTR language (e.g. an Israeli AA group, or one whose meeting list is in Arabic) might throw a spanner in the works.

**Reply by joshreisner** [reply]
Do you have a platform preference @lloy0076? I would probably go for Laravel with DomPDF but I don't know enough about what else is out there to know if that's the best choice. 

I think a good middle-of-the-road option to try first would be a single vertical fold down an 8.5x11" piece of paper, into two 4.25x11" columns. The rationale being that this is such a common paper size in the USA, and it would be such a simple fold to make for most people. 

What would take it to the next level would be to bookletize it so that when you print out 10 pages and fold down the middle and staple that the pages end up in proper order [like this](https://helpx.adobe.com/acrobat/kb/print-booklets-acrobat-reader.html).

**Reply by lloy0076** [reply]
@joshreisner I'd be comfortable with Laravel and an appropriate PDF library such as https://packagist.org/packages/barryvdh/laravel-dompdf (or just plain DomPDF itself).

A simple fold would be nice.

That said, it might be wise to force people to print to booklet because that makes the proper ordering and fold _someone else's problem_ and it reduces what we need to do.

Someone might have to do a simple "How To" video on how to print the outcome, though; if people could cope with my Australian accent -- and providing I was properly "anonymised" -- I could probably make such a video...

**Reply by lloy0076** [reply]
PS. On a pragmatic level, I'd be most efficient coding this up in Laravel and PHP; others here are clearly familiar with PHP.

I would personally steer away from anything Ruby on Rails (not because it's incapable, this is purely personal preference or prejudice) and although Mojolicious (or some other Perl framework) / Perl would be perfectly capable of doing what we need, I don't know if we'd have enough volunteers to help with the code.

I'm sure that something in the C#/.Net ecosystem could do this (ASP.Net.Core isn't bad these days) but that's my area of least expertise (on the other hand, it would be an interesting "resume driven development" exercise for me).

So that leaves the PHP frameworks such as Symfony or Laravel; or a server side JS implementation (on Node using Express / Hapi or something similar); and of course the more esoteric choices - although some would argue that Golang wouldn't be that esoteric.

Pragmatically I'd settle on Laravel or Symfony with a preference for Laravel.

**Reply by joshreisner** [reply]
Started a repo at https://github.com/code4recovery/pdf - @lloy0076 if you want to contribute please do

**Reply by lloy0076** [reply]
@joshreisner Yes, I've just noticed; I've started to fill out its Wiki a little bit; whilst I have your attention enabling a Discussion tab on it might be useful too.

It might be useful for us (or a group of us) to have some sort of conference about a roadmap for it and so forth :)

**Reply by joshreisner** [reply]
[test.pdf](https://github.com/code4recovery/12-step-meeting-list/files/6835070/test.pdf)

got a basic proof of concept working if you run it locally. i think we just need to get this published somewhere so people can try it out with their own JSON feeds.

**Reply by joshreisner** [reply]
better proof-of-concept. this time i prepended and appended two google docs for the front and back matter, the repo only prints the inside pages

[schedule.pdf](https://github.com/code4recovery/12-step-meeting-list/files/6835477/schedule.pdf)

not bad right? seems almost usable. the trick i think will be to figure out bookletized printing. should be possible with adobe reader though.

![Screen Shot 2021-07-17 at 5 22 00 PM](https://user-images.githubusercontent.com/1551689/126052237-dc3f133c-0325-4a77-8231-dde046a8c2c0.png)

**Reply by Grateful-Viv** [reply]
I’m excited to see this progress. Thanks @lloy0076 & @joshreisner for taking this on. Personally I find printed meeting schedules a waste of paper and time, but so many people love them.

**Reply by lloy0076** [reply]
Booklet publishing can be made **someone else's problem**; we could of course have the document(s) print out in the correct page order but so long as we print them in order, we can make getting printing the end users' issue...so if you printed that in booklet form you get:

[booklet-print.pdf](https://github.com/code4recovery/12-step-meeting-list/files/6835578/booklet-print.pdf)

Now, we could definitely figure out the order ourselves (there's various ways) but why make our lives difficult.

As far as I can tell, most (or any) Windows print drivers can print to booklet.

ALSO we could use `pdfbook` ourselves to do this (https://linux.die.net/man/1/pdfbook) depending on the service we are using [if we're using a cheap hosting option, maybe not, but if we have our own VM then we definitely could).

THIS is a version from `pdfbook` on Ubuntu 20.04:

[schedule-book.pdf](https://github.com/code4recovery/12-step-meeting-list/files/6835586/schedule-book.pdf)

**Reply by lloy0076** [reply]
@joshreisner Do you happen to know what this is:

- [https://github.com/code4recovery/booklet-generator](https://github.com/code4recovery/booklet-generator)

**Reply by lloy0076** [reply]
> @joshreisner Do you happen to know what this is:
> 
>     * https://github.com/code4recovery/booklet-generator

Ok, I installed that, the relevant react scripts and _amazingly_ it didn't crash whilst building.

Apart from annoying CORS errors, I'm hitting:

https://github.com/manuels/texlive.js/issues/43 (I think).

Does anyone know if that works/worked?

**Reply by lloy0076** [reply]
Ok, it DOES work after a bit of fiddling, and I get:

[booklet-generator.pdf](https://github.com/code4recovery/12-step-meeting-list/files/6835617/booklet-generator.pdf)

So, TL;DR it looks like we could even attempt to get the full power of TeX or LaTeX to run _in the browser_ (honestly, I've never heard of doing such an evil thing but there's nothing new under the sun).

Now, if this really *does* work, though...we could hijack that code and use it -- and if it behaves on most modern browsers, we'd be able to offload a bunch of functionality onto the end user's browser AND lose the need for a microservice.

How delightfully evil 👿

**Reply by lloy0076** [reply]
> 
> 
> > @joshreisner Do you happen to know what this is:
> > ```
> > * https://github.com/code4recovery/booklet-generator
> > ```
> 
> Ok, I installed that, the relevant react scripts and _amazingly_ it didn't crash whilst building.
> 
> Apart from annoying CORS errors, I'm hitting:
> 
> [manuels/texlive.js#43](https://github.com/manuels/texlive.js/issues/43) (I think).
> 
> Does anyone know if that works/worked?

I got this working by:

- Downloading one of District 42's JSON caches to my local web server;
- Fixing the CORS issue (very lazily, I might add - I just added a "Oh, who cares what origin it's from" header into the default Virtual Host - a bit of a 🔨 solution but it's a test local site);
- Changed the line endings of the `textlive.lst` to just `LF`

And voila!

**Reply by lloy0076** [reply]
Right so that builds the meetings up using LaTeX itself, e..g:

```
function singleMeetingToLatex(meeting, print_strings){
	let newname = latexEscapeCharacterFilter(meeting.name);
	let newaddress = latexEscapeCharacterFilter(meeting.address);
	let newnotes = latexEscapeCharacterFilter(meeting.notes);
	let newtime = formatTime(meeting.time, print_strings);

	//need to format time
	let latex = "\\settowidth{\\extSpace}{" + newtime + "} \\setlength{\\extSpace}{4.2em - \\extSpace} ";
		latex = latex + newtime + " \\hspace{\\extSpace} \\textbf{"+newname+"} \\hfill "+meeting.city+"\n";
		latex = latex + "\\settowidth{\\extSpace}{" + meeting.types + "} \\setlength{\\extSpace}{4.2em - \\extSpace} ";
		latex = latex + "\\\\ " + meeting.types + " \\hspace{\\extSpace} " + newaddress + newnotes;

	return latex;
}
```

I guess the disadvantage of reusing this is that the PDF generator isn't the same as the one used in the proposed PDF microservice; and LaTeX can be somewhat persnickety.

That said, LaTeX isn't that hard to understand or use...

**Reply by lloy0076** [reply]
> I’m excited to see this progress. Thanks @lloy0076 & @joshreisner for taking this on. Personally I find printed meeting schedules a waste of paper and time, but so many people love them.

Aye, this may be so but there are people who do prefer paper lists (and as I had mentioned above, they are still being used).

**Comment by alangrady** [comment]
> It appears that Chris T's plugin seems to work (give or take PHP warnings).

I Really wished it work. :(

I getting "There has been a critical error on this website."

It would be Great if we could have a way to printed out meeting schedules.

(Sorry my English is bad. I had a stroke last year) 🕺

**Reply by joshreisner** [reply]
Would you like to try upgrading to the `develop` branch of our plugin? We have some updates coming that should a) fix the existing PDF, and b) add a new PDF option. 

There is some risk involved. If you want to wait that would be the safest route. Important note: our new version requires PHP 5.6 or higher. Please confirm your PHP version before you do this.

If so try downloading this ZIP file https://github.com/code4recovery/12-step-meeting-list/archive/refs/heads/develop.zip and uploading it via FTP to your plugins directory or WP Dashboard > Plugins > Add New > Upload Plugin 🤞 hopefully everything will work.

This update doesn't make any changes to data, but it does update our PHP quite a bit. If something goes wrong you can roll back to the current version with https://github.com/code4recovery/12-step-meeting-list/archive/refs/heads/master.zip

In that event, having FTP access will be handy.

**Reply by joshreisner** [reply]
We've just begun testing this version, but so far we haven't found any issues on our demo site at https://demo.code4recovery.org/meetings

**Comment by alangrady** [comment]
I got "500 Server Error" when I hit "Generate PDF".  😱

**Reply by joshreisner** [reply]
Ok so it sounds like you were able to install the new version without incident which is good, then you click Generate PDF and you get a 500? Could you share with me the URL in your browser when you see the error?

When I click Generate PDF from the plugin from the demo site I'm not seeing an error, so sounds like something is different in your case that we need to account for

**Reply by Grateful-Viv** [reply]
@joshreisner, can you put a link on the demo site to see the PDF in action? 

Is it only selecting in-person meetings?

Thanks.

**Reply by joshreisner** [reply]
@Grateful-Viv sure it's https://pdf.code4recovery.org/?json=https%3A%2F%2Fdemo.code4recovery.org%2Fwp-admin%2Fadmin-ajax.php%3Faction%3Dmeetings

Yes, it's only selecting meetings that meet in person

**Comment by alangrady** [comment]
https://pdf.code4recovery.org/pdf?json=https%3A%2F%2Fraleighaa.com%2Fwp-admin%2Fadmin-ajax.php%3Faction%3Dmeetings&type=&width=4.25&height=11&numbering=1&language=en&font=sans-serif&mode=download <https://pdf.code4recovery.org/pdf?json=https://raleighaa.com/wp-admin/admin-ajax.php?action=meetings&type=&width=4.25&height=11&numbering=1&language=en&font=sans-serif&mode=download>

**Reply by joshreisner** [reply]
Great, thanks. Should be working now.

Also the normal PDF links (which it says are going away soon) should be working too, right?

**Comment by alangrady** [comment]
👍 🕺  Generated PDF is now working!!!!


The normal pdf links is still not wrapping the meeting types.

http://snpy.in/8tSd2j

**Reply by joshreisner** [reply]
Ok I'm glad the new one is working and the old one isn't throwing a critical error anymore.

The old one is getting retired we won't be making any improvements to it. Hopefully the new one will evolve to be a much better option. Please report any issues over at https://github.com/code4recovery/pdf/issues

**Comment by alangrady** [comment]
You ROCK!  🕺

**Comment by CathyN-Oklahoma** [comment]
Has there been any changes to the PDF formats available on the 12 step plug in?  or a separate plug in?  I find that I am unable to edit the pdf that is downloaded to add additional pages/information.  Is that something particular to me or everyone?

**Comment by anchovie91471** [comment]
Hey Cathy! There have been some minor changes. We changed the fonts we use to support Japanese characters and we've added a couple of things here or there, but nothing wholesale.

Are you using the instructions here for assembling a booklet? https://github.com/code4recovery/pdf

**Comment by CathyN-Oklahoma** [comment]
No!  But I am on it now!  I will let you know if I have any questions.
Thank you so much and for your quick response.

Cathy

**Comment by CathyN-Oklahoma** [comment]
Yes, I have looked and tried that method.  But what I would like to do is
add pages to the front and back but I can't edit the pdf.  It says Adobe
has encountered an unknown error.  I also would like different sort
options.  We had a solution that worked really well but the process stopped
when the website our friend was using improved security so he couldn't
import our database anymore.  I'd like to make into a booklet form instead
of pages stapled together.

Any hints or ideas welcomed!

Cathy

**Comment by anchovie91471** [comment]
Can you send me the link to your feed so I can generate one and take a look?

**Comment by CathyN-Oklahoma** [comment]
My friend came in and helped.  He showed me how to print in booklet style.
I can add a cover separately.  Thank you so much for helping me.  I think
mostly we have a user-failure.
Cathy

**Reply by anchovie91471** [reply]
Glad it worked out!
